src/native/clr/include/host/pinvoke-override-impl.hh

```diff
@@ -0,0 +1,177 @@
+		if (entry_handle == nullptr) {
+			log_warn (LOG_ASSEMBLY, "Symbol '{}' not found in shared library '{}', p/invoke may fail", library_name, symbol_name);
+			return nullptr;
+		}
+
+		return entry_handle;
+	}
+
+	PINVOKE_OVERRIDE_INLINE
+	auto PinvokeOverride::load_library_entry (std::string const& library_name, std::string const& entrypoint_name, pinvoke_api_map_ptr api_map) noexcept -> void*
+	{
+		// Make sure some other thread hasn't just added the entry
+		auto iter = api_map->find (entrypoint_name);
+		if (iter != api_map->end () && iter->second != nullptr) {
+			return iter->second;
+		}
+
+		void *entry_handle = load_library_symbol (library_name.c_str (), entrypoint_name.c_str ());
+		if (entry_handle == nullptr) {
+			// error already logged
+			return nullptr;
+		}
+
+		log_debug (LOG_ASSEMBLY, "Caching p/invoke entry {} @ {}", library_name, entrypoint_name);
+		(*api_map)[entrypoint_name] = entry_handle;
+		return entry_handle;
+	}
+
+	PINVOKE_OVERRIDE_INLINE
+	void PinvokeOverride::load_library_entry (std::string_view const& library_name, std::string_view const& entrypoint_name, PinvokeEntry &entry, void **dso_handle) noexcept
+	{
+		void *entry_handle = load_library_symbol (library_name, entrypoint_name, dso_handle);
+		void *expected_null = nullptr;
+
+		bool already_loaded = !__atomic_compare_exchange (
+			/* ptr */              &entry.func,
+			/* expected */         &expected_null,
+			/* desired */          &entry_handle,
+			/* weak */              false,
+			/* success_memorder */  __ATOMIC_ACQUIRE,
+			/* failure_memorder */  __ATOMIC_RELAXED
+		);
+
+		if (already_loaded) {
+			log_debug (LOG_ASSEMBLY, "Entry '{}' from library '{}' already loaded by another thread", entrypoint_name, library_name);
+		}
+	}
+
+	PINVOKE_OVERRIDE_INLINE
+	auto PinvokeOverride::fetch_or_create_pinvoke_map_entry (std::string const& library_name, std::string const& entrypoint_name, hash_t entrypoint_name_hash, pinvoke_api_map_ptr api_map, bool need_lock) noexcept -> void*
+	{
+		auto iter = api_map->find (entrypoint_name, entrypoint_name_hash);
+		if (iter != api_map->end () && iter->second != nullptr) {
+			return iter->second;
+		}
+
+		if (!need_lock) {
+			return load_library_entry (library_name, entrypoint_name, api_map);
+		}
+
+		StartupAwareLock lock (pinvoke_map_write_lock);
+		return load_library_entry (library_name, entrypoint_name, api_map);
+	}
+
+	PINVOKE_OVERRIDE_INLINE
+	auto PinvokeOverride::find_pinvoke_address (hash_t hash, const PinvokeEntry *entries, size_t entry_count) noexcept -> PinvokeEntry*
+	{
+		while (entry_count > 0uz) {
+			const size_t mid = entry_count / 2uz;
+			const PinvokeEntry *const ret = entries + mid;
+			const std::strong_ordering result = hash <=> ret->hash;
+
+			if (result < 0) {
+				entry_count = mid;
+			} else if (result > 0) {
+				entries = ret + 1;
+				entry_count -= mid + 1uz;
+			} else {
+				return const_cast<PinvokeEntry*>(ret);
+			}
+		}
+
+		return nullptr;
+	}
```

Comment:
since we are doing a binary search, does this imply that the hash is sorted? If that is the case you probably can replace the code with the std::lower_bound https://en.cppreference.com/w/cpp/algorithm/lower_bound.html

Maybe something like:
```cpp
auto it = std::lower_bound(
    entries, entries + entry_count, hash,
    [](const PinvokeEntry& entry, hash_t value) {
        return entry.hash < value;
    }
);
if (it != entries + entry_count && it->hash == hash) {
    return const_cast<PinvokeEntry*>(it);
}
return nullptr;
```